include/User.class.php

```diff
@@ -1545,6 +1545,170 @@ public static function validateHomepage($homepage)
+     *
+     * @param int $id
+     *
+     * @return player info with rank
+     */
+    public static function getRanking($id)
+    {
+        try
+        {
+            $result = DBConnection::get()->query(
+                "SELECT
+                    FIND_IN_SET(scores,
+                        (SELECT GROUP_CONCAT(DISTINCT scores ORDER BY scores DESC)
+                        FROM `{DB_VERSION}_rankings`)
+                    ) AS rank,
+                    id,
+                    username,
+                    scores,
+                    max_scores,
+                    num_races_done
+                FROM `{DB_VERSION}_rankings`
+                WHERE `id` = :id",
+                DBConnection::FETCH_FIRST,
+                [':id' => $id],
+                [':id' => DBConnection::PARAM_INT]
+            );
+        }
+        catch (DBException $e)
+        {
+            Debug::addException($e);
+            // Initial scores for all users:
+            return
+                [
+                    'rank' => -1,
+                    'scores' => 2000.0,
+                    'max_scores' => 2000.0,
+                    'num_races_done' => 0
+                ];
+        }
+        if (!$result)
+        {
+            return
+                [
+                    'rank' => -1,
+                    'scores' => 2000.0,
+                    'max_scores' => 2000.0,
+                    'num_races_done' => 0
+                ];
+        }
+        return $result;
+    }
+
+    /**
+     * Reset ranking of all players (called only by user with PERM_SUMBIT_RANKINGS)
+     *
+     * @param User $user
+     *
+     * @throws UserException
+     */
+    public static function resetRanking($permission)
+    {
+        if (!in_array(AccessControl::PERM_SUMBIT_RANKINGS, $permission))
+        {
+            throw new UserException("Invalid user to reset ranking", ErrorType::USER_INVALID_PERMISSION);
+        }
+        try
+        {
+            DBConnection::get()->query("DELETE FROM `{DB_VERSION}_rankings`");
+        }
+        catch (DBException $e)
+        {
+            Debug::addException($e);
+        }
+    }
+
+    /**
+     * Submit new ranking of a player (called only by user with PERM_SUMBIT_RANKINGS)
+     *
+     * @param User $user
+     * @param int $id_for_ranked
+     * @param double $new_scores
+     * @param double $new_max_scores
+     * @param int $new_num_races_done
+     *
+     * @throws UserException
+     */
+    public static function submitRanking($permission, $id_for_ranked, $new_scores, $new_max_scores, $new_num_races_done)
+    {
+        if (!in_array(AccessControl::PERM_SUMBIT_RANKINGS, $permission))
+        {
+            throw new UserException("Invalid user to sumbit ranking", ErrorType::USER_INVALID_PERMISSION);
+        }
+        try
+        {
+            DBConnection::get()->query(
+                "REPLACE INTO `{DB_VERSION}_rankings` (id, username, scores,
```

Comment:
`REPLACE INTO` Seems  to delete then insert if the primary keys exist. (+ does other bad performance stuff http://code.openark.org/blog/mysql/replace-into-think-twice)
Use instead `INSERT .. ON DUPLICATE` https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html